clouddriver-lambda/src/main/java/com/netflix/spinnaker/clouddriver/lambda/provider/agent/LambdaCachingAgent.java

```diff
@@ -128,52 +129,25 @@ public CacheResult loadData(ProviderCache providerCache) {
     long loadDataStart = clock.instant().toEpochMilli();
     log.info("Describing items in {}", getAgentType());
 
-    AWSLambda lambda = amazonClientProvider.getAmazonLambda(account, region);
+    Map<String, CacheData> lambdaCacheData = new ConcurrentHashMap<>();
+    Map<String, Collection<String>> appLambdaRelationships = new ConcurrentHashMap<>();
 
     // Get All Lambda's
-    List<FunctionConfiguration> lstFunction = getFreshLambdas(lambda);
-
-    Map<String, CacheData> lambdaCacheData = new HashMap<>();
-    Map<String, Collection<String>> appLambdaRelationships = new HashMap<>();
-    for (FunctionConfiguration x : lstFunction) {
-      String functionKey =
-          Keys.getLambdaFunctionKey(getAccountName(), getRegion(), x.getFunctionName());
-
-      // Call AWS with individual lambda and build attributes to obtain LambdaConfiguration.State
-      // See notes for ->
-      // https://docs.aws.amazon.com/cli/latest/reference/lambda/list-functions.html
-      Map<String, Object> lambdaAttributes = buildLambdaAttributes(x.getFunctionName(), lambda);
-      if (lambdaAttributes == null) {
-        // The lambda was deleted between the list call and now
-        continue;
-      }
-
-      /* TODO: If the functionName follows frigga by chance (i.e. somename-someothername), it will try to store the
-         lambda as a relationship with the app name (somename), even if it wasn't deployed by spinnaker!
-      */
-      // Add the spinnaker application relationship and store it
-      Names names = Names.parseName(x.getFunctionName());
-      if (names.getApp() != null) {
-        String appKey =
-            com.netflix.spinnaker.clouddriver.aws.data.Keys.getApplicationKey(names.getApp());
-        Collection<String> functionKeys =
-            appLambdaRelationships.getOrDefault(appKey, new ArrayList<>());
-        functionKeys.add(functionKey);
-        appLambdaRelationships.put(appKey, functionKeys);
-        lambdaCacheData.put(
-            functionKey,
-            new DefaultCacheData(
-                functionKey,
-                lambdaAttributes,
-                Collections.singletonMap(APPLICATIONS.ns, Collections.singletonList(appKey))));
-      } else {
-        // TODO: Do we care about non spinnaker deployed lambdas?
-        lambdaCacheData.put(
-            functionKey,
-            new DefaultCacheData(functionKey, lambdaAttributes, Collections.emptyMap()));
-      }
+    List<Map<String, Object>> allLambdas;
+    try {
+      allLambdas = lambdaService.getAllFunctions();
+    } catch (Exception e) {
+      throw new SpinnakerException(
+          "Failed to populate the lambda cache for account '"
+              + account.getName()
+              + "' and region '"
+              + region
+              + "' because: "
+              + e.getMessage());
     }
 
+    buildCacheData(lambdaCacheData, appLambdaRelationships, allLambdas);
```

Comment:
Didnt dig enough to know if it is possible or not, but do we need to build the cache data in one batch, or can we do it incrementally? like when we have complete data for some functions collected, we update the cache for those functions and iterate. It will add even more performance benefits if possible

Reply:
Same as above -- I think we can try to optimize a few areas. Will note this as well